baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ExceptionSpecificity.java

```diff
@@ -0,0 +1,415 @@
+                        getThrownCheckedExceptions(tree, state), state);
+                if (containsBroadException(thrownCheckedExceptions, state)) {
+                    return Description.NO_MATCH;
+                }
+                ImmutableList<Type> thrown = flattenExceptionTypes(thrownCheckedExceptions, state);
+                if (thrown.size() <= MAX_CHECKED_EXCEPTIONS
+                        // Do not apply this to test code where it's likely to be noisy.
+                        // In the future we may want to revisit this.
+                        && !TestCheckUtils.isTestCode(state)) {
+                    List<Type> replacements = deduplicateCatchTypes(
+                            ImmutableList.<Type>builder()
+                                    .addAll(thrown)
+                                    .addAll((isThrowable ? THROWABLE_REPLACEMENTS : EXCEPTION_REPLACEMENTS).stream()
+                                            .map(name -> Preconditions.checkNotNull(
+                                                    state.getTypeFromString(name), "Failed to find type"))
+                                            .collect(ImmutableList.toImmutableList()))
+                                    .build(),
+                            encounteredTypes,
+                            state);
+                    SuggestedFix.Builder fix = SuggestedFix.builder();
+                    if (replacements.isEmpty()) {
+                        // If the replacements list is empty, this catch block isn't reachable and can be removed.
+                        fix.replace(catchTree, "");
+                    } else {
+                        Name parameterName = catchTree.getParameter().getName();
+                        AssignmentScanner assignmentScanner = new AssignmentScanner(parameterName);
+                        catchTree.getBlock().accept(assignmentScanner, null);
+                        if (replacements.size() == 1 || !assignmentScanner.variableWasAssigned) {
+                            catchTree.accept(new ImpossibleConditionScanner(
+                                    fix, replacements, parameterName), state);
+                            fix.replace(catchTypeTree, replacements.stream()
+                                    .map(type -> SuggestedFixes.prettyType(state, fix, type))
+                                    .collect(Collectors.joining(" | ")));
+                        }
+                    }
+                    return buildDescription(catchTypeTree)
+                            .addFix(fix.build())
+                            .build();
+                }
+                return Description.NO_MATCH;
+            }
+            // mark the type as caught before continuing
+            encounteredTypes.add(catchType);
+        }
+        return Description.NO_MATCH;
+    }
+
+    @Override
+    @SuppressWarnings("CyclomaticComplexity")
+    public Description matchMethod(MethodTree tree, VisitorState state) {
```

Comment:
Wondering if it would be clearer if this lived in a separate class? That way we'd have two error-prone checks that just do one thing, e.g. CatchSpecificity and ThrowsSpecificity?

Reply:
I could go either way on this, there are two manifestations of the same conceptual issue. I suppose it boils down to whether we want bug checkers mapped to concepts, or implementation/output?

I was thinking about this yesterday, specifically for the assertj cleanup checkers -- it might be easiest for consumers to understand if all assertj refactor checks were part of a single 'AssertjUsage' check.

Reply:
I think I'm mainly thinking about this from a maintenance perspective, as shorter classes that just do one thing are easier to understand as I can be sure each private method is somehow relevant to the one purpose of the class.

Reply:
That's a good point